Add time since last weather update to weather app

JustScott opened the pull request:
Provides the time since the last weather update, incrementing the duration from seconds to minutes at 60 seconds, and minutes to hours at 3600 seconds. I'm open to suggestions on changes to the texts position and/or color. This should resolve #2140.

Also, on the simulator the seconds old count is only updated when you open the app, while on hardware the seconds count up live while in the app... I'm not sure why this is.

![weather_0s_old](https://github.com/user-attachments/assets/59937917-015b-489e-a7f8-bbaecff9a2b3)
![weather_32s_old](https://github.com/user-attachments/assets/69d68a2a-0062-4bad-86b3-f0d90ab01077)
![weather_1m_old](https://github.com/user-attachments/assets/0bafcfc5-5bed-4d42-b7c9-af025b78e8b6)

JustScott committed:
Add time since last weather update to weather app

liamcharger commented:
Personally, I would prefer to see the minutes instead of the seconds when I have recently updated my weather. For instance, instead of displaying “32 seconds old,” I would like to see “Now” and only show the number of minutes after one has passed.

Another aspect that feels more natural to me is using the phrase “X minutes ago” instead of “X minutes old.”

JustScott commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -137,6 +144,18 @@ void Weather::Refresh() {
+
+      int64_t secondsSinceEpoch = dateTimeController.CurrentDateTime().time_since_epoch().count() / 1000000000;
+      int64_t secondsSinceWeatherUpdate = secondsSinceEpoch - optCurrentWeather->timestamp;
+      if (secondsSinceWeatherUpdate < 0) {
```
I did have it showing -1 a few times in the simulator, so figured I would add that just to cover it in the off chance it happened on the watch.

JustScott commented:
> Personally, I would prefer to see the minutes instead of the seconds when I have recently updated my weather. For instance, instead of displaying “32 seconds old,” I would like to see “Now” and only show the number of minutes after one has passed.
> 
> Another aspect that feels more natural to me is using the phrase “X minutes ago” instead of “X minutes old.”

I agree with you on using `ago` instead of `old`. As for using `Now` instead of displaying the seconds, I feel this looks a little akward and personally prefer the seconds count, I'd like to hear what others think about this.


![now_instead_of_seconds](https://github.com/user-attachments/assets/938c89cc-18d5-48db-acd0-00e307e09d14)
![seconds_instead_of_now](https://github.com/user-attachments/assets/1de86c9b-c152-4a58-ad7b-4bc96b16f5ed)

Have to move the label up `-1px` to prevent the `g` from being cut off like in the picture above
![move_update_text_up](https://github.com/user-attachments/assets/5bd396a8-52ad-466a-812f-9188f4f4d447)

And I move the temp up as well to prevent crowding:
![move_all_up](https://github.com/user-attachments/assets/1cc0e89a-9557-4010-9076-0d0cdb1d6557)

JustScott committed:
Replace 'old' with 'ago' and adjust positioning

tituscmd commented:
I like `Now` as well, but also agree that it looks a bit awkward.
I have some propositions as to how this could be perhaps made better:

`Updated now` / `Updated 12s ago`
`Fetched now` / `Fetched 12s ago`
`Refreshed now` / `Refreshed 12s ago`

Adding another word results in a less awkward look, as well as being more direct and intuitive to understand.

tituscmd commented:
> ![seconds_instead_of_now](https://private-user-images.githubusercontent.com/68134791/409827721-1de86c9b-c152-4a58-ad7b-4bc96b16f5ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4Mzc5MjcsIm5iZiI6MTczODgzNzYyNywicGF0aCI6Ii82ODEzNDc5MS80MDk4Mjc3MjEtMWRlODZjOWItYzE1Mi00YTU4LWFkN2ItNGJjOTZiMTZmNWVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDEwMjcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRkMmNhZTllMWRlYTVmNjdkNGU2ODE0Njg3MjQ5MjVjOTRiODRkMWMxOTUxOWE2NTVkODRmNTJmYmJiMDgzODImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dyFWE-SEePcS23QnKOcwxuFfLCWqDYT_wXO2HehH1Os)
>
> ![test](https://private-user-images.githubusercontent.com/68134791/409828372-5bd396a8-52ad-466a-812f-9188f4f4d447.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg4Mzc5MjcsIm5iZiI6MTczODgzNzYyNywicGF0aCI6Ii82ODEzNDc5MS80MDk4MjgzNzItNWJkMzk2YTgtNTJhZC00NjZhLTgxMmYtOTE4OGY0ZjRkNDQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA2VDEwMjcwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkMWRkYTMwNjdhMDkxMDk4MzI0YmJmNTFjNDgxNjQ5ODIxNjA2Mjg0MDRjNTE5Mzg5YjkzZDA1MDA5NzdlN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SMIvW5MWwo46j_Z6ChFznN9uTORhK6NJq0f2mfELMwM)

In these two images the label doesn't seem to be centered. Is that just me?

JustScott commented:
@tituscmd I fixed the centering issue by dynamically changing the position depending on weather there is on or two digits. I just tried out a bunch of examples of how this could look, everyone should let me know which they prefer!

### With a seconds counter (with dynamic alignment)
![InfiniSim_2025-02-06_170016](https://github.com/user-attachments/assets/c0c22af9-0941-41bd-8bf4-8e6b7991e6cd)
![InfiniSim_2025-02-06_170034](https://github.com/user-attachments/assets/9904cd6e-f817-4297-b909-5ced849b22c6)

### Two version of up to date
![InfiniSim_2025-02-06_170114](https://github.com/user-attachments/assets/463f6950-b601-4d62-beec-f33260998b73)
![up_to_date](https://github.com/user-attachments/assets/2184b900-b3f1-413b-906a-f2c45b573e88)

### variations of "updated now"
![now_instead_of_seconds](https://github.com/user-attachments/assets/5e0892b2-81be-4a99-b6ba-644ae0aa8cea)
![InfiniSim_2025-02-06_170643](https://github.com/user-attachments/assets/83029acb-e86e-4299-9279-14516692d3a3)
![InfiniSim_2025-02-06_170704](https://github.com/user-attachments/assets/67f4b548-d593-40a8-b27f-2075826e7dd2)
![InfiniSim_2025-02-06_170812](https://github.com/user-attachments/assets/9acd0b9b-7d4f-4784-9ba5-6552eaef8d6b)

### "Updated:" with dynamic alignment
![InfiniSim_2025-02-06_171040](https://github.com/user-attachments/assets/f7c64e2d-5f45-4656-8997-f5268ff340f0)
![InfiniSim_2025-02-06_171427](https://github.com/user-attachments/assets/7e6debb0-9c45-4e74-b9e4-8bcc7f208a55)

I can make more examples if anyones curious. Just let me know. I personally prefer to just include the seconds count, as it may be useful to some and keeps the label consistent... also the live second count on hardware makes the app feel more lively.

tituscmd commented:
> ![InfiniSim_2025-02-06_170034](https://private-user-images.githubusercontent.com/68134791/410693633-9904cd6e-f817-4297-b909-5ced849b22c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg5MTQ1MzcsIm5iZiI6MTczODkxNDIzNywicGF0aCI6Ii82ODEzNDc5MS80MTA2OTM2MzMtOTkwNGNkNmUtZjgxNy00Mjk3LWI5MDktNWNlZDg0OWIyMmM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAyMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMjA3VDA3NDM1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFkZGFhMWUwNDQxZTAwZjkxNDRiMzFhZjZkNmEzYjNiOTU5Yzg4MjVhY2ZiMjE4YThkMGZhOTVjOTBhMmQ2ZmQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.Pt3YHNDVD1H09w9msDy9OieDzSdnUa4W0SDKq_Ev6Oc)

In this image, if you look at the weather icon up top it still doesn't look centered. I don't know if you're doing that already, but I would suggest aligning the label with a pointer to the weather icon or the day in the middle of the table, those always seem to be centered.

I may be a bit of a perfectionist ;D

tituscmd commented:
@JustScott Thank you so much for making all these options!

I think after looking at everything, I like `Now` and `45s ago` the most. The others seem to make the app look to crowded.

JustScott commented:
> @JustScott Thank you so much for making all these options!
> 
> I think after looking at everything, I like `Now` and `58s ago` / `53m ago` / `15h ago` the most. The others seem to make the app look too crowded.

I agree with you on this, but am curious as to how long you think the app should display`Now` before switching to `Xs ago`?

tituscmd commented:
> > In this image, if you look at the weather icon up top it still doesn't look centered. I would suggest aligning the label with a pointer to the weather icon or the day in the middle of the table, those always seem to be centered.
> > I may be a bit of a perfectionist ;D
> 
> Ah I see what you mean about it not looking centered, I was just trying to keep the space between the `1m` and `ago` above the second letter of whatever the third day is, so it doesn't look like the label is moving when it moves from `9m` to `10m`.

I see, that makes sense.
I'd prefer it centered, but I like the idea of making it look like the label isn't moving! :)

tituscmd commented:
> I agree with you on this, but am curious as to how long you think the app should display`Now` before switching to `Xs ago`?

Oh, right. What about at `30s ago`?

JustScott commented:
> > I agree with you on this, but am curious as to how long you think the app should display`Now` before switching to `Xs ago`?
> 
> Oh, right. What about at `30s ago`? Or maybe `31s ago`, so that the first 30 seconds are displayed as "Now" and the next 30 seconds are displayed as `Xs ago`

Okay, just so anyone reading this is caught up, here's what it'll look like:

![now](https://github.com/user-attachments/assets/baac4959-b74e-4b4f-8197-635e5a94df27)
![31s_ago](https://github.com/user-attachments/assets/56836a0e-6bf5-4590-b8b5-6ef26bdd6174)
> I'd prefer it centered, but I like the idea of making it look like the label isn't moving! :)

I personally prefer it to be centered over the third words second letter, so I say we see what the others think. I'm open to changing it if people agree it doesn't look centered.
![one_minute](https://github.com/user-attachments/assets/253568e8-363a-45ed-9bca-4ee7634896ef)

JustScott committed:
Updated labels, improved/simplified logic, and improved positioning

Use "Now" for the label until 31 seconds, then display the live time.

JustScott committed:
Fixed positioning

Move from `lv_obj_align` to `lv_obj_set_pos` to avoid weird positioning
errors when changing the position more than once.

liamcharger commented:
Just a summary of what I would expect to see—"Now" until a minute has passed, "Xmin ago" until an hour has passed, and so on. Support for "Xd ago" and "Xw ago" would also make sense so I'm not looking at "80h ago," or something similar in case my weather isn't updated for a while.

As far as positioning, I like the way the label looks in @JustScott's comment above.

Here's an idea for another PR: instead of squeezing everything onto one screen, what if it's separated into two that the user can scroll vertically—one for the current weather and one for the forecast? The forecast days could be laid out horizontally, and this would fix the text wrap issue in #2240 and would eliminate the need to pick an element to center on in this PR.

JustScott commented:
> Support for "Xd ago" and "Xw ago" would also make sense so I'm not looking at "80h ago," or something similar in case my weather isn't updated for a while.


I believe the weather service only keeps the weather data for 24 hours, so after that it wouldn't display anything until updated again.

liamcharger commented:
> I believe the weather service only keeps the weather data for 24 hours, so after that it wouldn't display anything until updated again.

👍 Just ignore that part then, haha

dariusarnold commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -137,6 +144,18 @@ void Weather::Refresh() {
       lv_label_set_text_fmt(minTemperature, "%d°", minTemp);
       lv_label_set_text_fmt(maxTemperature, "%d°", maxTemp);
+
+      int64_t secondsSinceEpoch = dateTimeController.CurrentDateTime().time_since_epoch().count() / 1000000000;
```
Sorry, I was not on the most recent commit, my bad.

dariusarnold commented on src/displayapp/screens/Weather.cpp:
```diff
@@ -137,10 +143,44 @@ void Weather::Refresh() {
+
+      lv_obj_set_pos(lastUpdated, X_SINGLE_DIGIT_POSITION, Y_POSITION);
+
+      if (hoursSinceWeatherUpdate > 0) {
```
The `secondsSinceWeatherUpdate > 9` condition in line171  will always be true due to the if from the previous line and could be removed.

The `secondsSinceWeatherUpdate < 60` should also be always true because otherwise the flow would go through the `minutesSinceWeatherUpdate > 0` branch, but you can consider keeping it to be explicit (the same is true for `minutesSinceWeatherUpdate < 60` in line 166). The `secondsSinceWeatherUpdate < 31` condition in line 175 could then be transformed to just an `else`, probably with a comment to tell the reader that this is the < 31 s case.

The final code with all suggestions applied could look like this:
```cpp
      if (hoursSinceWeatherUpdate > 0) {
        if (hoursSinceWeatherUpdate > 9) {
          lv_obj_set_pos(lastUpdated, X_TWO_DIGIT_POSITION, Y_POSITION);
        }
        lv_label_set_text_fmt(lastUpdated, "%dh ago", hoursSinceWeatherUpdate);
      } else if (minutesSinceWeatherUpdate > 0) {
        if (minutesSinceWeatherUpdate > 9) {
          lv_obj_set_pos(lastUpdated, X_TWO_DIGIT_POSITION, Y_POSITION);
        }
        lv_label_set_text_fmt(lastUpdated, "%dm ago", minutesSinceWeatherUpdate);
      } else if (secondsSinceWeatherUpdate > 30) {
          lv_obj_set_pos(lastUpdated, X_TWO_DIGIT_POSITION, Y_POSITION);
        lv_label_set_text_fmt(lastUpdated, "%ds ago", secondsSinceWeatherUpdate);
      } else {
        // Update time less than 30 s
        lv_obj_set_pos(lastUpdated, X_NOW_POSITION, Y_POSITION);
        lv_label_set_text_fmt(lastUpdated, "Now", secondsSinceWeatherUpdate);
      }
```

JustScott committed:
Use `auto` for cast to meet Coding convention

Thanks to @dariusarnold for catching this

JustScott committed:
Reduce unnecessary logic in positioning code

Thanks to @dariusarnold for suggesting these changes. There was no need
to define an upper bounds of 60 for changing to two digit positioning.

JustScott committed:
Fixed formatting

JustScott commented:
@dariusarnold Thank you for the review, I've applied all of your suggestions outside of the final `else` statement instead of the explicit `else if` as I prefer this over a comment. Thank you for helping! Also, I don't really use weather much anymore on my pinetime... so if you find any bugs with these changes please let me know!

liamcharger commented:
Little bump on this to the maintainers :) would like to see this in 1.17

mark9064 commented:
agree
